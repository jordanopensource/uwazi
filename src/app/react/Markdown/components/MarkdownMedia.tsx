/* eslint-disable max-lines */
/* eslint-disable max-statements */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, useRef, useEffect } from 'react';
import { FieldArrayWithId, useFieldArray, useForm } from 'react-hook-form';
import ReactPlayer from 'react-player';
import { Icon } from 'UI';
import { Translate } from 'app/I18N';
import { validMediaFile } from 'app/Metadata/helpers/validator';

interface MarkdownMediaProps {
  compact?: boolean;
  editing?: boolean;
  onTimeLinkAdded?: Function;
  config: string;
  type?: string;
}

interface TimeLink {
  timeHours: string;
  timeMinutes: string;
  timeSeconds: string;
  label: string;
}

const propsToConfig = (props: MarkdownMediaProps) => {
  const config = { url: '', options: { timelinks: {} } };

  let parsedProps: any = props.config.replace(/\(|\)/g, '').split(',');
  config.url = parsedProps.shift() as string;

  parsedProps = (parsedProps.join(',') || '{}').replace(/&quot;/g, '"');

  try {
    parsedProps = JSON.parse(parsedProps);
  } catch (error) {
    parsedProps = {};
  }

  config.options = parsedProps;

  return config;
};

const formatTimeLinks = (timelinks: any): TimeLink[] =>
  Object.keys(timelinks).map(key => {
    const timeLink = { timeHours: '00', timeMinutes: '00', timeSeconds: '00', label: '' };
    const splitTimes = key.split(':');
    if (splitTimes.length === 2) {
      [timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    } else {
      [timeLink.timeHours, timeLink.timeMinutes, timeLink.timeSeconds] = splitTimes;
    }
    timeLink.label = timelinks[key] as string;
    return timeLink;
  });

const MarkdownMedia = (props: MarkdownMediaProps) => {
  //Given the testing conditions, ref should be declared in this specific order.
  const containerRef = useRef<HTMLDivElement>(null);
  const playerRef = useRef<ReactPlayer>(null);

  const [newTimeline, setNewTimeline] = useState<TimeLink>({
    timeHours: '',
    timeMinutes: '',
    timeSeconds: '',
    label: '',
  });
  const [errorFlag, setErrorFlag] = useState(false);
  const { options } = propsToConfig(props);
  const originalTimelinks = formatTimeLinks(options?.timelinks || {});
  const [playingTimelinkIndex, setPlayingTimelinkIndex] = useState<number>(-1);
  const [isVideoPlaying, setVideoPlaying] = useState<boolean>(false);
  const [temporalResource, setTemporalResource] = useState<string>();
  const [mediaURL, setMediaURL] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const { control, register, getValues } = useForm<{ timelines: TimeLink[] }>({
    defaultValues: { timelines: originalTimelinks },
  });
  const { fields, remove, append } = useFieldArray<{ timelines: TimeLink[] }>({
    control,
    name: 'timelines',
  });

  const validMediaUrlRegExp =
    /(^(blob:)?https?:\/\/(?:www\.)?)[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]$/;

  const seekTo = (seconds: number) => {
    const playingStatus = isVideoPlaying;
    playerRef.current?.seekTo(seconds);
    setVideoPlaying(playingStatus);
  };

  const timeLinks = (_timelinks: any) => {
    const timelinks = _timelinks || {};
    return Object.keys(timelinks).map((timeKey, index) => {
      const linkIndex = index;
      const seconds = timeKey
        .split(':')
        .reverse()
        .reduce((_seconds, n, _index) => _seconds + parseInt(n, 10) * 60 ** _index, 0);

      let displayTime = timeKey;
      if (displayTime.split(':').length === 2) {
        displayTime = `00:${displayTime}`;
      }

      return (
        <div className="timelink" key={timeKey + index.toString()} title={timelinks[timeKey]}>
          <b
            className="timelink-icon"
            onClick={() => {
              seekTo(seconds);
              if (linkIndex === playingTimelinkIndex) {
                setVideoPlaying(false);
                setPlayingTimelinkIndex(-1);
              } else {
                setVideoPlaying(true);
                setPlayingTimelinkIndex(linkIndex);
              }
            }}
          >
            <Icon icon={linkIndex === playingTimelinkIndex ? 'pause' : 'play'} />
          </b>
          <div
            className="timelink-time-label"
            onClick={() => {
              seekTo(seconds);
              setPlayingTimelinkIndex(linkIndex);
            }}
          >
            <b>{displayTime}</b>
            <span>{timelinks[timeKey]}</span>
          </div>
        </div>
      );
    });
  };

  const updateParentForm = () => {
    if (props.onTimeLinkAdded) props.onTimeLinkAdded(getValues().timelines);
  };

  const appendTimelinkAndUpdateParent = (timelink?: TimeLink) => {
    const currentTimelink = timelink || newTimeline;
    append(currentTimelink);
    updateParentForm();
  };

  const renderNewTimeLinkForm = () => (
    <div className="new-timelink">
      <div className="timestamp">
        <input
          type="number"
          onChange={event => {
            const hours = parseInt(event.target.value || '0', 10);
            setNewTimeline({ ...newTimeline, timeHours: hours <= 0 ? '00' : hours.toString() });
          }}
          className="timestamp-hours"
          placeholder="00"
          value={newTimeline.timeHours}
          min="0"
        />
        <span className="seperator">:</span>
        <input
          type="number"
          onChange={event => {
            let minutes = parseInt(event.target.value || '0', 10);
            minutes = minutes > 59 ? 59 : minutes;
            setNewTimeline({
              ...newTimeline,
              timeMinutes: minutes <= 0 ? '00' : minutes.toString(),
            });
          }}
          className="timestamp-minutes"
          placeholder="00"
          value={newTimeline.timeMinutes}
          min="0"
          max="59"
        />
        <span className="seperator">:</span>
        <input
          type="number"
          onChange={event => {
            let seconds = parseInt(event.target.value || '0', 10);
            seconds = seconds > 59 ? 59 : seconds;
            setNewTimeline({
              ...newTimeline,
              timeSeconds: seconds <= 0 ? '00' : seconds.toString(),
            });
          }}
          className="timestamp-seconds"
          placeholder="00"
          value={newTimeline.timeSeconds}
          min="0"
          max="59"
        />
      </div>
      <input
        type="text"
        onChange={event => {
          setNewTimeline({ ...newTimeline, label: event.target.value });
        }}
        className="timestamp-label"
        placeholder="Enter title"
        value={newTimeline.label}
      />
      <button
        type="button"
        className="new-timestamp-btn"
        onClick={() => {
          const time = {
            timeHours: newTimeline.timeHours || '00',
            timeMinutes: newTimeline.timeMinutes || '00',
            timeSeconds: newTimeline.timeSeconds || '00',
            label: newTimeline.label,
          };

          appendTimelinkAndUpdateParent(time);
          setNewTimeline({ timeHours: '', timeMinutes: '', timeSeconds: '', label: '' });
        }}
      >
        <Icon icon="plus" />
      </button>
    </div>
  );

  const renderSingleTimeLinkInputs = (field: FieldArrayWithId<TimeLink>, index: number) => (
    <div className="new-timelink" key={index}>
      <div className="timestamp">
        <input
          type="text"
          className="timestamp-hours"
          placeholder="00"
          key={`${field.id}hours`}
          {...register(`timelines.${index}.timeHours`, {
            onChange: _ => updateParentForm(),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-minutes"
          placeholder="00"
          key={`${field.id}minutes`}
          {...register(`timelines.${index}.timeMinutes`, {
            onChange: _ => updateParentForm(),
          })}
        />
        <span className="seperator">:</span>
        <input
          type="text"
          className="timestamp-seconds"
          placeholder="00"
          key={`${field.id}seconds`}
          {...register(`timelines.${index}.timeSeconds`, {
            onChange: _ => updateParentForm(),
          })}
        />
      </div>
      <input
        type="text"
        className="timestamp-label"
        placeholder="Enter title"
        key={field.id}
        {...register(`timelines.${index}.label`, {
          onChange: _ => updateParentForm(),
        })}
      />
      <button
        title="Remove timelink"
        type="button"
        className="delete-timestamp-btn"
        onClick={() => {
          remove(index);
          updateParentForm();
        }}
      >
        <Icon icon="trash-alt" />
      </button>
    </div>
  );

  const renderTimeLinksForm = () => (
    <>
      {fields.map((field: FieldArrayWithId<TimeLink>, index: number) =>
        renderSingleTimeLinkInputs(field, index)
      )}
      {renderNewTimeLinkForm()}
    </>
  );

  const config = propsToConfig(props);

  useEffect(() => {
    let url: string;

    if (config.url.startsWith('/api/files/')) {
      fetch(config.url)
        .then(async res => {
          if (validMediaFile(res)) {
            return res.blob();
          }
          setErrorFlag(true);
          throw new Error('Invalid file');
        })
        .then(blob => {
          setErrorFlag(false);
          url = URL.createObjectURL(blob);
          setMediaURL(url);
        })
        .catch(_e => {})
        .finally(() => {
          setIsLoading(false);
        });
    } else if (config.url.match(validMediaUrlRegExp)) {
      setErrorFlag(false);
      setMediaURL(config.url);
      setIsLoading(false);
    } else {
      if (mediaURL && mediaURL.match(validMediaUrlRegExp) && !temporalResource) {
        setTemporalResource(mediaURL);
      }
      setMediaURL(config.url);
      setIsLoading(false);
    }

    return () => {
      setErrorFlag(false);
      setIsLoading(true);
      URL.revokeObjectURL(url);
      setMediaURL('');
    };
  }, [config.url]);

  useEffect(() => () => {
    if (isVideoPlaying) {
      setVideoPlaying(false);
    }
  });

  useEffect(() => {
    if (
      temporalResource &&
      ReactPlayer.canPlay(temporalResource) &&
      !mediaURL.match(validMediaUrlRegExp)
    ) {
      setErrorFlag(false);
      setMediaURL(temporalResource);
    }
  }, [temporalResource, mediaURL]);

  const { compact, editing } = props;
  const dimensions: { width: string; height?: string } = { width: '100%' };
  if (compact) {
    dimensions.height = '100%';
  }

  if (errorFlag) {
    return (
      <div className="media-error">
        <Translate>This file type is not supported on media fields</Translate>
      </div>
    );
  }

  return (
    <div className={`video-container ${compact ? 'compact' : ''}`} ref={containerRef}>
      <div>
        {isLoading ? (
          <div className="loader">
            <Translate>Loading</Translate>
            <div className="bouncing-dots">
              <div className="dot" />
              <div className="dot" />
              <div className="dot" />
            </div>
          </div>
        ) : (
          <ReactPlayer
            className="react-player"
            playing={isVideoPlaying}
            config={{
              facebook: {
                attributes: {
                  'data-width': editing ? '' : containerRef.current?.clientWidth,
                  'data-height': editing ? '300' : containerRef.current?.clientHeight,
                },
              },
            }}
            ref={playerRef}
            url={mediaURL}
            {...dimensions}
            controls
            onPause={() => {
              setVideoPlaying(false);
            }}
            onPlay={() => {
              setVideoPlaying(true);
            }}
            onError={e => {
              if (e.target.error.message.search(/MEDIA_ELEMENT_ERROR/) === -1) {
                setErrorFlag(true);
              }
            }}
          />
        )}
      </div>
      {!editing && !isLoading && <div>{timeLinks(config.options.timelinks)}</div>}
      {editing && !isLoading && (
        <div className="timelinks-form">
          <button
            type="button"
            className="add-timelink"
            onClick={() => {
              const currentTime = playerRef.current?.getCurrentTime() as number;
              const hours = Math.floor(currentTime / 3600);
              const remainingSeconds = currentTime - hours * 3600;
              const minutes = Math.floor(remainingSeconds / 60);
              const seconds = Math.floor(remainingSeconds % 60);
              const timelink = {
                timeHours: hours < 10 ? `0${hours.toString()}` : hours.toString(),
                timeMinutes: minutes < 10 ? `0${minutes.toString()}` : minutes.toString(),
                timeSeconds: seconds < 10 ? `0${seconds.toString()}` : seconds.toString(),
                label: '',
              };
              appendTimelinkAndUpdateParent(timelink);
            }}
          >
            <Translate>Add timelink</Translate>
          </button>
          <h5>
            <Translate>Timelinks</Translate>
          </h5>
          {renderTimeLinksForm()}
        </div>
      )}
      <p className="print-view-alt">{config.url}</p>
    </div>
  );
};

export type { TimeLink, MarkdownMediaProps };
export default MarkdownMedia;
