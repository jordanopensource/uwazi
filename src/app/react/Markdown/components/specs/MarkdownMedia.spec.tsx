/**
 * @jest-environment jsdom
 */
import React from 'react';
import { RenderResult, fireEvent, screen, act } from '@testing-library/react';
import fetchMock from 'fetch-mock';

import { defaultState, renderConnectedContainer } from 'app/utils/test/renderConnected';
import MarkdownMedia from '../MarkdownMedia';

let playerRef: { current: { seekTo: any; getCurrentTime: () => number } };

const mockUseRef = jest.fn().mockImplementation(args => {
  const actual = jest.requireActual<typeof React>('react')
    ? jest.requireActual<typeof React>('react').useRef(args)
    : null;
  if (args === null && actual) {
    playerRef = actual;
    return actual;
  }
  return actual;
});

jest.mock('react', () => ({
  ...jest.requireActual<typeof React>('react'),
  useRef: (args: any) => mockUseRef(args),
}));

describe('MarkdownMedia', () => {
  let renderResult: RenderResult;
  const mockedCreateObjectURL: jest.Mock = jest.fn();
  const mockedRevokeObjectURL: jest.Mock = jest.fn();

  beforeEach(() => {
    URL.createObjectURL = mockedCreateObjectURL;
    mockedCreateObjectURL.mockReturnValue('blob:abc');
    URL.revokeObjectURL = mockedRevokeObjectURL;
    fetchMock.mock('/api/files/1683080859038pwqi670wk7r.mp4', {}, { overwriteRoutes: true });
    fetchMock.mock(
      'https://www.vimeo.com/253530307',
      { throws: 'CORS error' },
      { overwriteRoutes: true }
    );
  });

  afterEach(() => {
    mockedCreateObjectURL.mockReset();
    mockedRevokeObjectURL.mockReset();
  });

  const onTimeLinkAdded = jest.fn();

  const render = async (
    options: { compact?: boolean; editing?: boolean } = { compact: false, editing: false },
    URL = '/api/files/1683080859038pwqi670wk7r.mp4'
  ) => {
    const props = {
      ...options,
      onTimeLinkAdded,
      config: `${URL}, {"timelinks": {"02:10": "A rude awakening", "05:30": "Finally, you are up!"}})`,
    };
    await act(() => {
      ({ renderResult } = renderConnectedContainer(
        <MarkdownMedia {...props} />,
        () => defaultState
      ));
    });
  };

  describe('render', () => {
    describe('uploaded files', () => {
      it('should render an iframe that displays the video from the blob resource', async () => {
        await render();

        expect(mockedCreateObjectURL.mock.calls[0].toString()).toEqual('[object Blob]');

        expect(renderResult.asFragment()).toMatchSnapshot();
      });

      it('should revoke the created blob', async () => {
        await act(async () => {
          await render();
        });

        expect(mockedCreateObjectURL.mock.calls[0].toString()).toEqual('[object Blob]');

        await act(() => {
          renderResult.unmount();
        });

        expect(mockedRevokeObjectURL).toHaveBeenCalledWith('blob:abc');
      });

      it('should render the edition mode', async () => {
        await render({ editing: true });
        expect(renderResult.asFragment()).toMatchSnapshot();
      });
      it('should use compact class name if compact prop is set', async () => {
        await render({ compact: true });
        expect(renderResult.asFragment()).toMatchSnapshot();
      });
    });
    describe('external URLs', () => {
      it('should render an iframe with and point to the external resource', async () => {
        await render({ compact: false, editing: false }, 'https://www.vimeo.com/253530307');
        expect(renderResult.asFragment()).toMatchSnapshot();
        expect(mockedCreateObjectURL).not.toBeCalled();
      });
    });

    describe('Video timeline', () => {
      it('should render the timelinks', async () => {
        await render();
        const links = renderResult.container.getElementsByClassName('timelink');
        expect(links.length).toBe(2);
      });

      it('should interact with the player through time line label', async () => {
        await render();
        const spySeekTo = jest.spyOn(playerRef.current, 'seekTo');
        const firstTimeLink = screen.getByText('A rude awakening').parentElement!;

        await act(async () => {
          fireEvent.click(firstTimeLink);
          expect(spySeekTo).toHaveBeenCalledWith(2 * 60 + 10);
        });
        const secondTimeLink = screen.getByText('Finally, you are up!').parentElement!;

        await act(async () => {
          fireEvent.click(secondTimeLink);
          expect(spySeekTo).toHaveBeenCalledWith(5 * 60 + 30);
        });
      });

      const playTimeLink = async (
        index: number,
        duration: number,
        status: string,
        spySeekTo: jest.SpyInstance
      ) => {
        const timeLink = renderResult.container.getElementsByClassName('timelink-icon')[index];
        await act(async () => {
          fireEvent.click(timeLink);
          expect(spySeekTo).toHaveBeenCalledWith(duration);
          const icon = renderResult.container.getElementsByClassName('timelink-icon')[index];
          expect(icon.children[0].getAttribute('data-icon')).toEqual(status);
        });
      };
      it('should interact with the player through time line icon', async () => {
        await render();
        const spySeekTo = jest.spyOn(playerRef.current, 'seekTo');
        await playTimeLink(0, 2 * 60 + 10, 'play', spySeekTo);
        await playTimeLink(1, 5 * 60 + 30, 'play', spySeekTo);
        await playTimeLink(1, 5 * 60 + 30, 'pause', spySeekTo);
      });
      const editTimeLink = async (inputs: HTMLElement[], field: string, value: string) => {
        await act(async () => {
          fireEvent.change(inputs.find(x => (x as HTMLInputElement).name === field)!, {
            target: { value },
          });
        });
      };

      it('should allow to add a timelink', async () => {
        await render({ editing: true });
        jest.spyOn(playerRef.current, 'getCurrentTime').mockReturnValue(27784);
        const addLinkBtn = screen.getByText('Add timelink').parentElement!;
        await act(async () => {
          fireEvent.click(addLinkBtn);
        });
        await act(async () => {
          await editTimeLink(
            screen.getAllByRole('textbox'),
            'timelines.2.label',
            'Check at this!'
          )!;
          expect(onTimeLinkAdded).toHaveBeenCalledWith([
            { label: 'A rude awakening', timeHours: '00', timeMinutes: '02', timeSeconds: '10' },
            {
              label: 'Finally, you are up!',
              timeHours: '00',
              timeMinutes: '05',
              timeSeconds: '30',
            },
            { label: 'Check at this!', timeHours: '07', timeMinutes: '43', timeSeconds: '04' },
          ]);
        });
      });

      it('should allow to edit a saved timelink', async () => {
        await render({ editing: true });
        const removeTimeLinkButtons = screen.getAllByRole('button', { name: 'Remove timelink' });

        await act(async () => {
          fireEvent.click(removeTimeLinkButtons[0]);
        });
        await act(async () => {
          const inputs = screen.getAllByRole('textbox');
          await editTimeLink(inputs, 'timelines.0.timeHours', '04');
          await editTimeLink(inputs, 'timelines.0.timeMinutes', '42');
          await editTimeLink(inputs, 'timelines.0.timeSeconds', '56');
        });
        await act(async () => {
          expect(onTimeLinkAdded).toHaveBeenCalledWith([
            {
              label: 'Finally, you are up!',
              timeHours: '04',
              timeMinutes: '42',
              timeSeconds: '56',
            },
          ]);
        });
      });

      it('should allow to edit a new timelink', async () => {
        await render({ editing: true });
        const addLinkBtn = screen.getByText('Add timelink').parentElement!;
        await act(async () => {
          fireEvent.click(addLinkBtn);
        });
        const updatedInputs = screen.getAllByRole('textbox');

        await editTimeLink(updatedInputs, 'timelines.2.timeHours', '12');
        await editTimeLink(updatedInputs, 'timelines.2.timeMinutes', '17');
        await editTimeLink(updatedInputs, 'timelines.2.timeSeconds', '34');
        await editTimeLink(updatedInputs, 'timelines.2.label', 'new check point');
        expect(onTimeLinkAdded).toHaveBeenCalledWith([
          { label: 'A rude awakening', timeHours: '00', timeMinutes: '02', timeSeconds: '10' },
          {
            label: 'Finally, you are up!',
            timeHours: '00',
            timeMinutes: '05',
            timeSeconds: '30',
          },
          { label: 'new check point', timeHours: '12', timeMinutes: '17', timeSeconds: '34' },
        ]);
      });

      it('should update the default new timelink', async () => {
        await render({ editing: true });

        fireEvent.change(renderResult.container.getElementsByClassName('timestamp-hours')[2], {
          target: { value: 11 },
        });
        fireEvent.change(renderResult.container.getElementsByClassName('timestamp-minutes')[2], {
          target: { value: 36 },
        });
        fireEvent.change(renderResult.container.getElementsByClassName('timestamp-seconds')[2], {
          target: { value: 45 },
        });
        fireEvent.change(renderResult.container.getElementsByClassName('timestamp-label')[2], {
          target: { value: 'default check point' },
        });

        fireEvent.click(renderResult.container.getElementsByClassName('new-timestamp-btn')[0]);
        expect(onTimeLinkAdded).toHaveBeenCalledWith([
          { label: 'A rude awakening', timeHours: '00', timeMinutes: '02', timeSeconds: '10' },
          {
            label: 'Finally, you are up!',
            timeHours: '00',
            timeMinutes: '05',
            timeSeconds: '30',
          },
          {
            label: 'default check point',
            timeHours: '11',
            timeMinutes: '36',
            timeSeconds: '45',
          },
        ]);
      });
    });
  });
});
