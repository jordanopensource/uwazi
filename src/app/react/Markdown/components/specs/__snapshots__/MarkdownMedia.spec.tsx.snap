// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownMedia render external URLs should render an iframe with and point to the external resource 1`] = `
<DocumentFragment>
  <div
    class="video-container "
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 360px;"
      >
        <div
          style="width: 100%; height: 100%; overflow: hidden;"
        />
      </div>
    </div>
    <div>
      <div
        class="timelink"
        title="A rude awakening"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:02:10
          </b>
          <span>
            A rude awakening
          </span>
        </div>
      </div>
      <div
        class="timelink"
        title="Finally, you are up!"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:05:30
          </b>
          <span>
            Finally, you are up!
          </span>
        </div>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      https://www.vimeo.com/253530307
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkdownMedia render uploaded files should render an iframe that displays the video from the blob resource 1`] = `
<DocumentFragment>
  <div
    class="video-container "
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 360px;"
      >
        <video
          controls=""
          preload="auto"
          src="blob:abc"
          style="width: 100%; height: 100%;"
        />
      </div>
    </div>
    <div>
      <div
        class="timelink"
        title="A rude awakening"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:02:10
          </b>
          <span>
            A rude awakening
          </span>
        </div>
      </div>
      <div
        class="timelink"
        title="Finally, you are up!"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:05:30
          </b>
          <span>
            Finally, you are up!
          </span>
        </div>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      /api/files/1683080859038pwqi670wk7r.mp4
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkdownMedia render uploaded files should render the edition mode 1`] = `
<DocumentFragment>
  <div
    class="video-container "
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 360px;"
      >
        <video
          controls=""
          preload="auto"
          src="blob:abc"
          style="width: 100%; height: 100%;"
        />
      </div>
    </div>
    <div
      class="timelinks-form"
    >
      <button
        class="add-timelink"
        type="button"
      >
        <span
          class="translation active"
        >
          Add timelink
        </span>
      </button>
      <h5>
        <span
          class="translation active"
        >
          Timelinks
        </span>
      </h5>
      <div
        class="new-timelink"
      >
        <div
          class="timestamp"
        >
          <input
            class="timestamp-hours"
            name="timelines.0.timeHours"
            placeholder="00"
            type="text"
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-minutes"
            name="timelines.0.timeMinutes"
            placeholder="00"
            type="text"
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-seconds"
            name="timelines.0.timeSeconds"
            placeholder="00"
            type="text"
          />
        </div>
        <input
          class="timestamp-label"
          name="timelines.0.label"
          placeholder="Enter title"
          type="text"
        />
        <button
          class="delete-timestamp-btn"
          title="Remove timelink"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-trash-alt fa-w-14 "
            data-icon="trash-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="new-timelink"
      >
        <div
          class="timestamp"
        >
          <input
            class="timestamp-hours"
            name="timelines.1.timeHours"
            placeholder="00"
            type="text"
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-minutes"
            name="timelines.1.timeMinutes"
            placeholder="00"
            type="text"
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-seconds"
            name="timelines.1.timeSeconds"
            placeholder="00"
            type="text"
          />
        </div>
        <input
          class="timestamp-label"
          name="timelines.1.label"
          placeholder="Enter title"
          type="text"
        />
        <button
          class="delete-timestamp-btn"
          title="Remove timelink"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-trash-alt fa-w-14 "
            data-icon="trash-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="new-timelink"
      >
        <div
          class="timestamp"
        >
          <input
            class="timestamp-hours"
            min="0"
            placeholder="00"
            type="number"
            value=""
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-minutes"
            max="59"
            min="0"
            placeholder="00"
            type="number"
            value=""
          />
          <span
            class="seperator"
          >
            :
          </span>
          <input
            class="timestamp-seconds"
            max="59"
            min="0"
            placeholder="00"
            type="number"
            value=""
          />
        </div>
        <input
          class="timestamp-label"
          placeholder="Enter title"
          type="text"
          value=""
        />
        <button
          class="new-timestamp-btn"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      /api/files/1683080859038pwqi670wk7r.mp4
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkdownMedia render uploaded files should use compact class name if compact prop is set 1`] = `
<DocumentFragment>
  <div
    class="video-container compact"
  >
    <div>
      <div
        class="react-player"
        style="width: 100%; height: 100%;"
      >
        <video
          controls=""
          preload="auto"
          src="blob:abc"
          style="width: 100%; height: 100%;"
        />
      </div>
    </div>
    <div>
      <div
        class="timelink"
        title="A rude awakening"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:02:10
          </b>
          <span>
            A rude awakening
          </span>
        </div>
      </div>
      <div
        class="timelink"
        title="Finally, you are up!"
      >
        <b
          class="timelink-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-play fa-w-14 "
            data-icon="play"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
              fill="currentColor"
            />
          </svg>
        </b>
        <div
          class="timelink-time-label"
        >
          <b>
            00:05:30
          </b>
          <span>
            Finally, you are up!
          </span>
        </div>
      </div>
    </div>
    <p
      class="print-view-alt"
    >
      /api/files/1683080859038pwqi670wk7r.mp4
    </p>
  </div>
</DocumentFragment>
`;
